File: 0494-target-sum/0494-target-sum.java
class Solution {

    private int [] nums;
    private int target;
    private int totalSum;
    private int[][] dp;
    private int n;

    public int findTargetSumWays(int[] nums, int target) {
        this.nums = nums;
        this.target = target;    

        n = nums.length;

        totalSum = 0;
        for (int i = 0; i < n; i++) {
            totalSum += nums[i];
        }

        dp = new int [n][2 * totalSum + 1];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }

        return findTotalWays(0, 0);
    }

    private int findTotalWays(int i, int sum) {
        
        if (sum == target && i == n) {
            return 1;
        }

        if (i >= n) {
            return 0;
        }

        if (dp[i][sum + totalSum] != -1) {
            return dp[i][sum + totalSum];
        }

        return dp[i][sum + totalSum]
             = findTotalWays(i + 1, sum + nums[i])
             + findTotalWays(i + 1, sum - nums[i]);
    }
}

File: 0494-target-sum/0494-target-sum.py
from typing import List

class Solution:
    def __init__(self):
        self.nums: List[int] = []
        self.target: int = 0
        self.totalSum: int = 0
        self.dp: List[List[int]] = []
        self.n: int = 0

    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        self.nums = nums
        self.target = target
        self.n = len(nums)

        self.totalSum = sum(nums)

        # Initialize dp with -1
        self.dp = [[-1] * (2 * self.totalSum + 1) for _ in range(self.n)]

        return self._findTotalWays(0, 0)

    def _findTotalWays(self, i: int, curr_sum: int) -> int:
        if curr_sum == self.target and i == self.n:
            return 1

        if i >= self.n:
            return 0

        if self.dp[i][curr_sum + self.totalSum] != -1:
            return self.dp[i][curr_sum + self.totalSum]

        add = self._findTotalWays(i + 1, curr_sum + self.nums[i])
        subtract = self._findTotalWays(i + 1, curr_sum - self.nums[i])

        self.dp[i][curr_sum + self.totalSum] = add + subtract
        return self.dp[i][curr_sum + self.totalSum]
